catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/OperationsCrudSupport.java

```diff
@@ -0,0 +1,503 @@
+                                    attributeName,
+                                    defaultValue)));
+                });
+    }
+
+    void generateMetacardAndContentItems(StorageRequest storageRequest,
+            List<ContentItem> incomingContentItems, Map<String, Metacard> metacardMap,
+            List<ContentItem> contentItems, Map<String, Path> tmpContentPaths)
+            throws IngestException {
+        for (ContentItem contentItem : incomingContentItems) {
+            try {
+                Path tmpPath = null;
+                long size;
+                try (InputStream inputStream = contentItem.getInputStream()) {
+                    if (inputStream == null) {
+                        throw new IngestException(
+                                "Could not copy bytes of content message.  Message was NULL.");
+                    }
+
+                    String sanitizedFilename =
+                            InputValidation.sanitizeFilename(contentItem.getFilename());
+                    tmpPath = Files.createTempFile(FilenameUtils.getBaseName(sanitizedFilename),
+                            FilenameUtils.getExtension(sanitizedFilename));
+                    Files.copy(inputStream, tmpPath, StandardCopyOption.REPLACE_EXISTING);
+                    size = Files.size(tmpPath);
+                    tmpContentPaths.put(contentItem.getId(), tmpPath);
+                } catch (IOException e) {
+                    if (tmpPath != null) {
+                        FileUtils.deleteQuietly(tmpPath.toFile());
+                    }
+                    throw new IngestException("Could not copy bytes of content message.", e);
+                }
+                String mimeTypeRaw = contentItem.getMimeTypeRawData();
+                mimeTypeRaw = guessMimeType(mimeTypeRaw, contentItem.getFilename(), tmpPath);
+
+                if (!InputValidation.checkForClientSideVulnerableMimeType(mimeTypeRaw)) {
+                    throw new IngestException("Unsupported mime type.");
+                }
+
+                String fileName = updateFileExtension(mimeTypeRaw, contentItem.getFilename());
+                Metacard metacard = generateMetacard(mimeTypeRaw,
+                        contentItem.getId(),
+                        fileName,
+                        (Subject) storageRequest.getProperties()
+                                .get(SecurityConstants.SECURITY_SUBJECT),
+                        tmpPath);
+                metacardMap.put(metacard.getId(), metacard);
+
+                ContentItem generatedContentItem = new ContentItemImpl(metacard.getId(),
+                        com.google.common.io.Files.asByteSource(tmpPath.toFile()),
+                        mimeTypeRaw,
+                        fileName,
+                        size,
+                        metacard);
+                contentItems.add(generatedContentItem);
+            } catch (Exception e) {
```

Comment:
We are using IngestException to break out of the loop then wrapping it in another IngestException just to do cleanup, their has to be a better way.